File: opendc-simulator/opendc-simulator-compute/src/main/java/org/opendc/simulator/compute/cpu/CpuPowerModel.java
/*
 * Copyright (c) 2022 AtLarge Research
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.opendc.simulator.compute.cpu;

import org.opendc.simulator.compute.machine.SimMachine;

/**
 * A model for estimating the power usage of a {@link SimMachine} based on the CPU usage.
 */
public interface CpuPowerModel {
    /**
     * Computes CPU power consumption for each host.
     *
     * @param utilization The CPU utilization percentage.
     * @return A double value of CPU power consumption (in W).
     */
    double computePower(double utilization);

    String getName();

    default String getFullName() {
        return getName();
    }
}

File: simulator/opendc-compute/opendc-compute-simulator/src/main/kotlin/org/opendc/compute/simulator/power/api/CpuPowerModel.kt
package org.opendc.compute.simulator.power.api

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import org.opendc.simulator.compute.SimMachine

public interface CpuPowerModel {
    /**
     * Computes CPU power consumption for each host.
     *
     * @param cpuUtil The CPU utilization percentage.
     * @return A [Double] value of CPU power consumption.
     * @throws IllegalArgumentException Will throw an error if [cpuUtil] is out of range.
     */
    @Throws(IllegalArgumentException::class)
    public fun computeCpuPower(cpuUtil: Double): Double

    /**
     * Emits the values of power consumption for servers.
     *
     * @param machine The [SimMachine] that the model is measuring.
     * @return A [Flow] of values representing the server power draw.
     */
    public fun getPowerDraw(machine: SimMachine): Flow<Double> =
        machine.usage.map {
            computeCpuPower(it)
        }
}
